### PR TITLE
Update Host, Via, and other headers in-place when possible

### PR DESCRIPTION
**Pull request by wweric**
This change optimizes Host header synchronization code that usually does
not need to change the Host header field but was always deleting the old
field and generating/appending a new one. Appending Host also violated
an RFC 9110 section 7.2 rule (on behalf of the user agent whose Host we
were rewriting): "A user agent that sends Host SHOULD send it as the
first field in the header section".

Also, removing old header fields and appending new ones introduced HTTP
header changes that broke otherwise working peers. For example, some
origins denied Squid-proxied messages exclusively due to an unusual
(from those origins point of view) Host header position. This change
reduces (but does not eliminate) such unnecessary and risky changes.

### CODE REVIEW

**Comment by rousskov** [comment]
OK to test

File: src/HttpHeader.cc [hunk 1]
```diff
@@ -1114,6 +1111,43 @@ HttpHeader::putExt(const char *name, const char *value)
...
 }
 
+void
+HttpHeader::updateOrAddStr(const Http::HdrType id, const SBuf &newValue)
+{
+    assert(any_registered_header(id));
+    assert(Http::HeaderLookupTable.lookup(id).type == Http::HdrFieldType::ftStr);
+
+    // XXX: HttpHeaderEntry::value suffers from String size limits
+    Must(newValue.length() < String::SizeMaxXXX()); // TODO: Assure() in master/v6
```
Comment (yadij) [inline src/HttpHeader.cc:1121]: Now this is against master the TODO can be implemented.

**Review by yadij — APPROVED** [review]
No objections from me.

The TODO that could be implemented I do not insist on.

**Comment by wweric** [comment]
@rousskov, regarding the build issue noted before with master branch source:

I'm seeing issue in the build logs from https://github.com/squid-cache/squid/suites/9708653393/artifacts/465427395 in any of the 4 log files. Search for FATAL, case sensitive. Trying to build locally I'm having issue along the same lines which it appears that most of the binaries/scripts that populate the /libexec folder are not being built at all during 'make'.  

The master branch source is only one I've had this problem with, any ideas?

File: src/HttpHeader.cc (hunk 1 above)
Comment (rousskov) [reply src/HttpHeader.cc:1121]: @wweric, I cannot implement this TODO because this PR is using a branch that is protected in your repository (or, at the very least, that is how I interpret what GitHub tells me when I try to edit this file). Please 
* replace this specific Must() with Assure();
* `#include "base/Assure.h"` (immediately after "squid.h") at the top of this .cc file; and
* remove the TODO comment on this line. The XXX comment above it should stay.

P.S. FWIW, it is best to use dedicated, PR-specific branches for PRs, but there is no need to adjust this PR branch.

File: src/HttpHeader.h [hunk 2]
```diff
@@ -142,6 +142,11 @@ class HttpHeader
     void putRange(const HttpHdrRange * range);
     void putSc(HttpHdrSc *sc);
     void putExt(const char *name, const char *value);
+
+    /// Ensures that the header has the given field, removing or replacing any
+    /// same-name fields with conflicting values as needed.
+    void updateOrAddStr(Http::HdrType, const SBuf &);
+    
     int getInt(Http::HdrType id) const;
```
[the remark this answers is not in the document]
Comment (wweric) [reply src/HttpHeader.h]: I just committed, but wondering if it's not the exact same, reinserted exactly what it took out in same spot...

[the remark this answers is not in the document]
Comment (rousskov) [reply src/HttpHeader.h]: You can see the trailing space being removed by your commit at https://github.com/squid-cache/squid/pull/1203/commits/2f05d74898c250fa4bfcdf9776686f33c9cae54a.

**Comment by rousskov** [comment]
> I'm seeing issue in the build logs from https://github.com/squid-cache/squid/suites/9708653393/artifacts/465427395 in any of the 4 log files. Search for FATAL, case sensitive.

```
2022/12/06 22:55:31| FATAL: MIME Config Table /home/runner/work/squid/squid/btlayer-01-minimal/squid-6.0.0-VCS/_inst/etc/mime.conf: (2) No such file or directory
2022/12/06 22:55:31| FATAL: logfile_daemon /home/runner/work/squid/squid/btlayer-01-minimal/squid-6.0.0-VCS/_inst/libexec/log_file_daemon: (2) No such file or directory
2022/12/06 22:55:31| FATAL: Icon Directory /home/runner/work/squid/squid/btlayer-01-minimal/squid-6.0.0-VCS/_inst/share/icons: (2) No such file or directory
```

Those FATAL messages are misleading: The missing file is not actually treated as a FATAL problem during that specific test -- see requirePathnameExists(). Misleading text aside, they are probably an indication a test-builds.sh bug (including bugs in the files that script is using). I did not check the details, but I suspect that the tests require that Squid is _installed_ but they are executed _before_ Squid is installed. Here is an abridged btlayer-01-minimal.log illustrating the suspected problem:

```shell
...
make  all-recursive # Squid is built but not installed
...
make  check-TESTS # some of these tests require Squid to be installed
...
make  install-recursive # ... but Squid is only installed later
```

FWIW, I get those FATAL messages when running test-builds.sh as well. However, I either did not notice them before (because the tests succeed despite them) or decided that they are an old known problem (because the test log is usually full of various error messages). Noisy test logs is one of the annoying existing testing deficiencies that nobody has fixed yet...

In summary, the false FATAL messages in CI build logs probably do not reflect true build problems in your environment.


> Trying to build locally I'm having issue along the same lines which it appears that most of the binaries/scripts that populate the /libexec folder are not being built at all during 'make'.

Perhaps you are missing some dependencies that are required to enable those binaries/scripts? If you are running a Linux distribution that uses apt, look for "build-dep" in squid/.github/workflows/default.yaml for one way to install most of the optional dependencies.

File: src/HttpHeader.cc [hunk 3]
```diff
@@ -991,10 +991,7 @@ HttpHeader::addVia(const AnyP::ProtocolVersion &ver, const HttpHeader *from)
         if (!strVia.isEmpty())
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/HttpHeader.cc:991]: Done at [02d77f7](https://github.com/squid-cache/squid/pull/1203/commits/02d77f7b893577087e42eb0bc7b17e594da731cd). Resolving.

File: src/HttpHeader.cc (hunk 1 above)
Comment (rousskov) [reply src/HttpHeader.cc:1121]: @wweric, just to avoid misunderstanding: AFAICT, this is the last change request that should be addressed before this PR is merged.

**Comment by wweric** [comment]
@rousskov, Thank You for allowing me to contribute to this project!

**Comment by rousskov** [comment]
> > I'm seeing issue in the build logs from https://github.com/squid-cache/squid/suites/9708653393/artifacts/465427395 in any of the 4 log files. Search for FATAL, case sensitive.
> 
> ```
> 2022/12/06 22:55:31| FATAL: MIME Config Table /home/runner/work/squid/squid/btlayer-01-minimal/squid-6.0.0-VCS/_inst/etc/mime.conf: (2) No such file or directory
> 2022/12/06 22:55:31| FATAL: logfile_daemon /home/runner/work/squid/squid/btlayer-01-minimal/squid-6.0.0-VCS/_inst/libexec/log_file_daemon: (2) No such file or directory
> 2022/12/06 22:55:31| FATAL: Icon Directory /home/runner/work/squid/squid/btlayer-01-minimal/squid-6.0.0-VCS/_inst/share/icons: (2) No such file or directory
> ```
> 
> Those FATAL messages are misleading: The missing file is not actually treated as a FATAL problem during that specific test -- see requirePathnameExists(). Misleading text aside, they are probably an indication a test-builds.sh bug (including bugs in the files that script is using). I did not check the details, but I suspect that the tests require that Squid is _installed_ but they are executed _before_ Squid is installed. Here is an abridged btlayer-01-minimal.log illustrating the suspected problem:
> 
> ```shell
> ...
> make  all-recursive # Squid is built but not installed
> ...
> make  check-TESTS # some of these tests require Squid to be installed
> ...
> make  install-recursive # ... but Squid is only installed later
> ```

I hope #1204 improves this situation, but this PR does not depend on that improvement.

**Comment by rousskov** [comment]
> @rousskov, I'm not sure if made a mistake or not by making the last two commits

To be more precise, you made a mistake much earlier -- when using your local "master" branch for an official PR branch ;-).

Now, AFAICT, your best option is to undo or revert the last out-of-scope commit (a96d031a28e29e9601e074210103e074186ab5e3) and wait. Once this PR merges and closes, then you can untangle your master branch from this PR, making your master branch a true copy of the official master branch. The latter will require you to force-push to your master, but that is OK in this special case.

We cannot merge this PR until that out-of-scope change in commit a96d031a28e29e9601e074210103e074186ab5e3 is removed. I will wait for that to happen before approving GitHub Actions tests, stalling this PR.

You can also revert/undo the merge commit (2fc6f45) but that should not introduce any changes to this PR so, AFAICT, it should not be a problem. Anubis will squash all PR commits when merging this PR.

For future PRs, see our [GitHints](https://wiki.squid-cache.org/GitHints), and do not hesitate to ask questions!

**Comment by wweric** [comment]
@rousskov, I just reverted a96d031 and Thank You for the pointer as I need to come up to speed with Git as never used it before and learning on the fly Vs actually reading and experimenting with my own repository.

So I'm thinking so far that basically if the web interface for example is used to fork Squid as I did, although many commits can be made, but only one PR at a time?

**Comment by rousskov** [comment]
> @rousskov, I just reverted [a96d031](https://github.com/squid-cache/squid/commit/a96d031a28e29e9601e074210103e074186ab5e3)

Thank you. If there are no more surprises, Anubis will wait for all the tests/timeouts to end, stage your changes, wait for all the tests against those staged changes to complete, and then merge the PR into the official "master" branch. Please do not modify your master branch in any way until this PR is closed (or we ask you for more changes).

> So I'm thinking so far that basically if the web interface for example is used to fork Squid as I did, although many commits can be made, but only one PR at a time?

No, you are still thinking too much about modifying official/master branches, which is something you never do (directly). Instead, try to follow this sketch:

* Configure your git client to work with two repositories -- official one and yours.
* Create a dedicated branch for each feature/project you want to work on. This feature branch will always live in your git repository.
* If you are _not_ going to create internal/unofficial PRs in your own GitHub space, against your own repository, then leave those cloned official branches in your repository "as is". Never modify them. You can even delete them! They are not really needed in this scenario AFAICT. Your feature branch will usually start from one of the official branches in the official repository. Your official PRs (like this one), will propose to merge your feature branch changes into one of the official branches (usually "master", "main", or whatever the project is using for their "active development" branch).
* If you do plan to create internal/unofficial PRs in your own GitHub space, then update cloned official branches in your repository to keep them in sync with the official repository. Your feature branch will usually start from one of the cloned official branches in your repository. Your unofficial PRs (in your GitHub space), will propose to merge your changes into one of the cloned (and synced) official branches in your repository. That way, you can test and polish your changes using GitHub tools _before_ you propose your changes for the official inclusion here. Your official PRs (like this one), will propose to merge your changes into one of the official branches (usually "master"). The underlying branch will be the same for internal and official PRs -- a feature branch in your repository.

**Comment by wweric** [comment]
@rousskov, IC now, I think, Thank you!

Fork a master clone that stays synced of course, then in effort to keep the master clone untouched, can branch a 'staging' branch(synced) below the master clone from which other branches can be created for individual PR's to be pushed to 'staging' branch first for triggering tests, polishing, or whatever before making an official PR.

Something like that, lol?

I'm working on getting it figured out...

**Comment by rousskov** [comment]
> Fork a master clone that stays synced of course,

Technically, one usually clones an entire official _repository_, not just one official master branch. The official repository has many branches, including, for example, the official v4 and v5 branches.

> then in effort to keep the master clone untouched, can branch a 'staging' branch(synced) below the master clone from which other branches can be created for individual PR's to be pushed to 'staging' branch first for triggering tests, polishing, or whatever before making an official PR.

There is an unnecessary extra branch in your description:

* I would avoid the term "staging". It is used for a different purpose (i.e. for checking what would happen if a PR branch is merged into the target branch of that PR without actually updating that target branch). Remove all the words about that staging branch from your description.
* Start your feature branches off the freshly synced master branch. Which repository that master branch is located is unimportant as far as this branching is concerned. Your feature branches will be in your repository, of course.
* Use your feature branches to post PRs (internal or official) against the master branch (cloned or official).
* Never merge such internal PRs -- to avoid confusion, at least when you are just starting to work with git, Squid master branch (in any repository) should only be changed by the Squid Project, via official PRs. If you need to accumulate or group your changes (from multiple feature branches), create another branch(es) for that accumulation (e.g., my-v5-based-changes or my-master-based-changes or whatever non-standard name makes sense to you).

Specific recipes are available at [GitHints](https://wiki.squid-cache.org/GitHints).

**Comment by wweric** [comment]
Ok, I was actually wondering about the extra 'staging' branch in the middle idea and whether or not it was a good idea/needed, Thank You.
